grants_tagger/scispacy_meshtagger.py

```diff
@@ -0,0 +1,59 @@
+"""
+Wrapper for scispacy mesh tagger. SciSpacy implements mesh
+tagging as a named entity recognition and linking step so
+the class also transforms the problem to multi label classification
+"""
+from collections import defaultdict
+
+from scispacy.linking import EntityLinker
+from scispacy.abbreviation import AbbreviationDetector
+from sklearn.metrics import f1_score
+from scipy.sparse import csr_matrix
+import scispacy
+import spacy
+
+
+class SciSpacyMeshTagger():
+    def __init__(self, mesh_tags, threshold=2):
+        """
+        mesh_tags: list of mesh UI descriptros e.g. [D00001]
+        """
+        self.mesh_tags = mesh_tags
+        self.threshold = threshold
+
+    def fit(self, *_):
+        nlp = spacy.load("en_core_sci_sm")
+        abbreviation_resolver = AbbreviationDetector(nlp)
+        nlp.add_pipe(abbreviation_resolver)
+        linker = EntityLinker(resolve_abbreviations=True, name="mesh")
+        nlp.add_pipe(linker)
+        self.nlp = nlp
+
+    def _get_tags(self, doc):
+        mesh_ents = defaultdict(int)
+        for ent in doc.ents:
+            if ent._.kb_ents and ent._.kb_ents[0][0] in self.mesh_tags:
+                mesh_ent = ent._.kb_ents[0]
+                mesh_ent_ui = mesh_ent[0]
+                mesh_ent_prob = mesh_ent[1]
+                mesh_ents[mesh_ent_ui] += mesh_ent_prob
```

Comment:
I wonder whether you mean if the probability of this mesh term is >0.5 this will be adding 1 and otherwise will be adding 0 (due to `mesh_ents = defaultdict(int)`)? If this is the case I think you need a comment to clear it up

Reply:
or even 
```
if mesh_ent_prob > 0.5:
   mesh_ents[mesh_ent_ui] += 1

Reply:
this is probably a bug, will investigate. well spotted

Reply:
Turns out that python comes to the rescue and converts to float. Int is used only to initialise in case no key is found.
```
>>> from collections import defaultdict
>>> a = defaultdict(int)
>>> a['b']
0
>>> a['b'] += .08
>>> a
defaultdict(<class 'int'>, {'b': 0.08})
```

Reply:
Slighly more information here https://stackoverflow.com/questions/5900578/how-does-collections-defaultdict-work. The argument passed, in this case int, is used to initialise the elements that are not present. Then normal python takes over and when you add a float to an int, python converts the int to float which what happens here

Reply:
ok great , good to know!